central/pruning/pruning.go

```diff
@@ -862,53 +867,48 @@ func (g *garbageCollectorImpl) collectAlerts(config *storage.PrivateConfig) {
 		return
 	}
 
-	ctx, cancel := context.WithTimeout(pruningCtx, alertQueryTimeout)
-	defer cancel()
-
 	prunedAlertCount := 0
 	// Go through the various queries and remove those batches.
 	for key, query := range queryMap {
 		log.Debugf("Pruning for key %q", key)
-		var alertsToPrune []string
-		// The alert searcher is opinionated and adds some default query parameters.
-		// Those should not be included for pruning.  Simpler to just use `WalkByQuery`
-		err := g.alerts.WalkByQuery(ctx, query, func(alert *storage.Alert) error {
-			alertsToPrune = append(alertsToPrune, alert.GetId())
-
-			return nil
-		})
+		alertsToPrune, err := g.getAlertsToPrune(query)
 		if err != nil {
 			log.Errorf("Unable to prune alerts for query %v", query)
 			continue
 		}
 
-		localBatchSize := alertDeleteBatchSize
-
-		for {
-			if len(alertsToPrune) == 0 {
-				break
-			}
-
-			if len(alertsToPrune) < localBatchSize {
-				localBatchSize = len(alertsToPrune)
-			}
+		func() {
+			pruneCtxWithTimeout, pruneCancel := contextutil.ContextWithTimeoutIfNotExists(pruningCtx, pruningTimeout)
+			defer pruneCancel()
 
-			alertBatch := alertsToPrune[:localBatchSize]
-			log.Infof("[Alert pruning] Removing %d alerts for %q", len(alertBatch), key)
-			if err := g.alerts.DeleteAlerts(pruningCtx, alertBatch...); err != nil {
+			log.Infof("[Alert pruning] Removing %d alerts for %q", len(alertsToPrune), key)
+			if err := g.alerts.PruneAlerts(pruneCtxWithTimeout, alertsToPrune...); err != nil {
 				log.Error(err)
 			} else {
-				prunedAlertCount = prunedAlertCount + len(alertBatch)
+				prunedAlertCount = prunedAlertCount + len(alertsToPrune)
 			}
-
-			// Move the slice forward to start the next batch
-			alertsToPrune = alertsToPrune[localBatchSize:]
-		}
+		}()
 	}
 
 	log.Infof("Pruned %d alerts based on retention configuration", prunedAlertCount)
 }
 
+func (g *garbageCollectorImpl) getAlertsToPrune(query *v1.Query) ([]string, error) {
+	var alertsToPrune []string
+	ctx, queryCancel := context.WithTimeout(pruningCtx, alertQueryTimeout)
+	defer queryCancel()
+
+	// The alert searcher is opinionated and adds some default query parameters.
+	// Those should not be included for pruning.  Simpler to just use `WalkByQuery`
+	err := g.alerts.WalkByQuery(ctx, query, func(alert *storage.Alert) error {
+		alertsToPrune = append(alertsToPrune, alert.GetId())
+
+		return nil
+	})
+
+	return alertsToPrune, err
```

Comment:
The alertsToPrune can be nonempty in case of err != nil. I saw the usage of getAlertsToPrune which did not handle the alerts ids in case of err.

Reply:
I updated that so we still try to process those in the event of an error here.  pruning some is better than pruning none